### PR TITLE
[SYCL] Make `handler.hpp` independent from `kernel_bundle.hpp`

### PR DESCRIPTION
**Pull request by AlexeySachkov**
I was looking into ways of splitting `sycl.hpp` into different finer-grained headers (with the intention to propose such split as a KHR extension/SYCL-Next thing) and I decided to try and see what is the impact of different header files on the compilation time.

I started my investigation with `kernel_bundle.hpp`. Looking at [zjin-lcf/HeCBench](https://github.com/zjin-lcf/HeCBench), I do not see any benchmarks that use it, so it seems like a good candidate for being an opt-in header.

To do the measurements I decided to drop `#include <kernel_bundle.hpp>` from `sycl.hpp` and then compare compilation time of two empty files including `sycl.hpp` (the modified one and the original one). Apparently, it is not that easy to drop an include, because there are so many inter-dependencies on it. I succeeded and I see ~200ms device compilation time improvement when `kernel_bundle.hpp` is not included at all.

However, for my experiments I made some other hacks which I'm unable to push into the repo, like dropping backend-specific headers as well. Specifically, L0 backend interop has `kernel_bundle` as a struct member of some of input or return types which requires a full definition.
SYCL spec ([6.3.7. Adding a backend](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_adding_a_backend)) allows backend interop headers to be put into separate headers and I think that we should actually use this opportunity in the future and drop them (somehow without many regressions) from `sycl.hpp`.

### CODE REVIEW

File: sycl/test/include_deps/sycl_detail_core.hpp.cpp
Comment (aelovikov-intel) [inline sycl/test/include_deps/sycl_detail_core.hpp.cpp:163]: Any removals here are good and desirable, I think :)

Also, feel free to add any new tests here (e.g. for `kernel_bundle.hpp` itself) if you'd find that beneficial to highlight the improvements (either this PR or any other of the same spirit).

File: sycl/include/sycl/handler.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/handler.hpp:1736]: I've done the same in a few places already for previous splits. I think mine are in `queue.hpp`

File: sycl/test/self-contained-handler-hpp.cpp
Comment (aelovikov-intel) [inline sycl/test/self-contained-handler-hpp.cpp:6]: Can we have tighter integration with your generic tests for self-contained-headers? Something like not generating generic test for a header if it has a dedicated specific one (like this).

File: sycl/include/sycl/ext/oneapi/backend/level_zero.hpp
Comment (AlexeySachkov) [reply sycl/include/sycl/ext/oneapi/backend/level_zero.hpp:44]: As I was working on #16030 and putting even more forward-declarations everywhere, I though that having a dedicated version of every header which only provides forward declarations would actually be cool.

Because some classes are templated, their forward-declaration may not be trivial and it may require more forward-declarations. Having a ready-to-use "canonical" forward-declaration in form of a separate header would shorten the codebase and make it simpler (i.e. that would just be a `#include` list without any extra lines of code).

There is a thing that some headers provide multiple classes and functions and therefore we could "import" more forward-declarations than we need, but I assume that it would be cheaper anyway that including the original header, so I don't think that it is a strong enough point against the idea.

File: sycl/test/self-contained-handler-hpp.cpp
Comment (AlexeySachkov) [reply sycl/test/self-contained-handler-hpp.cpp]: So, I did a thing! See 0d071507f097441485e846636d1d853d5359f23d

Apparently, the whole `self-contained-headers` suite has been disabled since #14879 and it fails. Therefore, I will outline this commit into a separate PR, fixing all the issues we have along the way. But it is still published here as a backup and to get early feedback

Comment (AlexeySachkov) [reply sycl/test/self-contained-handler-hpp.cpp]: Outlined this upgrade into #16117

**Review by EwanC — APPROVED** [review]
`sycl/test-e2e/Graph/graph_common.hpp` LGTM

**Review by YuriPlyakhin — APPROVED** [review]
Joint Matrix changes LGTM

File: sycl/test-e2e/syclcompat/math/math_compare.cpp
Comment (Alcpz) [inline sycl/test-e2e/syclcompat/math/math_compare.cpp:39]: I've checked, and this include shouldn't be needed. Could you please check-again? Otherwise, SYCLcompat changes LGTM

**Review by bader — APPROVED** [review]
👍

**Review by Alcpz — APPROVED** [review]
:+1:

**Review by aelovikov-intel — APPROVED** [review]

**Comment by AlexeySachkov** [comment]
I see some unexpected failures during local merge with `sycl` branch, in case some gatekeeper would want to merge the PR: please don't do that just yet. I will merge it myself after I re-check post-merge results.

**Review by sarnex — CHANGES_REQUESTED** [review]
i guess i can do this to prevent accidental automerge, ping me/gatekeepers when the PR is ready

**Comment by AlexeySachkov** [comment]
> i guess i can do this to prevent accidental automerge, ping me/gatekeepers when the PR is ready

Failures seem to be a side-effect of #15905 and unrelated to this PR, I can reproduce them on pure `sycl` branch as well. We can proceed with merge here, @sarnex

**Review by sarnex — APPROVED** [review]

**Pull request merged**
